### ISSUE TITLE
Async load collection for editors

### ISSUE THREAD

**Issue by asdman384** [ghiscoding/Angular-Slickgrid#39]
Hello,
I have next workflow:
Getting data for collection is deferred. How can I organize properly load of grid?

next tries didn't work: (I tried to re-init grid after filling collection)

```
 this.dataService.getStatuses().subscribe(data => {
       this.columns[0].editor.collection = data.map(item => ({ label: item.desc, value: item.id }))
       // this.grid.gridService.resetGrid(this.columns);
       this.grid.gridService.init(this.grid.slickGrid, this.dataset);
 }, this.handleError);
```

**Comment by ghiscoding** [comment]
Ah no the `init` function on each Services is to be used by the Service not really by the end user. 

Is it all the data that is deferred and all available at the same time? If so, just set the `dataset` at the time you are ready and the grid will refreshes. 

I think you are also trying to re-populate an `editor` but unfortunately the current implementation does not have any asynchronous functionality. You must pre-populate the `editor` and `filter` when creating the grid options.

**Comment by ghiscoding** [comment]
There is currently no way to load async collections for Editors & Filters. This is a limitation from SlickGrid itself, it would be nice to have this feature in Angular-Slickgrid in the future, but it is low priority since you can use an Angular Resolver to do the job.

So for a quick workaround, please use Angular Resolver to pre-populate your collections (before creating the grid).

**Comment by sabeurch** [comment]
Hi All, 
FOR FILTERS:
I managed doing this in a cutomized filter instance: the idea is:
1- Make reference to my master component (that holds http requests) from the custom filter.
The filter could accept as argument the master component.
```
export class SwtColumnFilter implements Filter {
   constructor(commonGrid:SwtCommonGrid )..
}
```

2- Make reference to my filter from the master component (this is easy as we have the instance of the filter being created when defining columns).

3- Now the filter's init function could be called from my custom component, 
```
init(args: FilterArguments) {
....
}
```
"args" is passed as NULL when being called from custom component, but being filled when it is called from Angular slick grid.
I am saving the "args" being first passed the angular slic grid, to reuse them once again and again.
```
    init(args: FilterArguments) {
      if(args == null && this.localArgs != null){
          args = this.localArgs;
      }
      else{
       ....
      }
  }
```

As for the data being used to construct the filter DOM, I am using thsi function from the custom component (that is called from the custom filtrer):

 ```
   getFilteredItems(): any{
        if(this.dataviewObj){
            let filtered: any = this.dataviewObj.getFilteredItems();
            return filtered != null&&filtered.length>0 ?filtered:this.dataset;
        }
        return this.dataset;
    }
```
FOR THE EDITORS:
It is planned in our requirements, but I am sure it will be based on custom editors since we are able to refer to our custom components from the editor itself (being injected in options for example).
 
Hope that can help.

**Comment by ghiscoding** [comment]
@sabeurch I'm not sure what the code you shown is doing. Is that for supporting async http for filter?

**Comment by sabeurch** [comment]
Yes, updating the filter content at each http request/reply that updates the grid dataset.

**Comment by ghiscoding** [comment]
@sabeurch 
is there something you can do in the lib to make async filter to work? or is that only working when creating a custom filter? I'm not sure the DI would work correctly in the lib, but I could be wrong

**Comment by ghiscoding** [comment]
@asdman384 @YadavAvinashK
I started some work to address this, it's a work in progress but I have the proof of concept working (not entirely tested though). You can provide feedback now before I push the code. I currently only worked on the Filter but the Editor will behave exactly the same.

#### Step 1 - use a new `asyncCollection` property
I created a new property named `asyncCollection` which accepts an Observable/Subject, so typically you would simply pass an HttpClient Observable. Also note that when you use `asyncCollection`, the other `collection` property can be omitted (however note the library itself will still use the `collection` property). So the code could look like the following
```javascript
this.columnDefinitions = [
{ 
  id: 'duration', name: 'Duration', field: 'duration', 
  filter: {
    asyncCollection: this.http.get('/api/durations'),
    model: Filters.multipleSelect,

    // add a collection sort if you wish, for demo let's sort DESC
    collectionSortBy: {
        property: 'value',
        sortDesc: true,
        fieldType: FieldType.number
    },
  }
}
```

#### Step 2 - how to refresh the Filter / Editor collection?
If you want to refresh or change the collection used by the Filter or Editor, you can still use the `collection` which will always contain the latest array used by the Filter/Editor (basically every time the async resolves, the lib updates the `collection` property). However, there's a slight problem, I could not find a way to simply watch the `collection` (like `$watchCollection` in AngularJS), so in order for the lib to know there's a change, the only way I found is by triggering a `next` on the Observable. So the code looks like the following
```javascript
  addItem() {
    // add a new row to the grid
    const lastRowIndex = this.dataset.length;
    const newRows = this.mockData(1, lastRowIndex);
    this.angularGrid.gridService.addItemToDatagrid(newRows[0]);

    // refresh the MultipleSelect Filter of the "Duration" column
    const durationColumnDef = this.columnDefinitions.find((column: Column) => column.id === 'duration');
    if (durationColumnDef) {
      const asyncCollection = durationColumnDef.filter.asyncCollection;
      const collection = durationColumnDef.filter.collection;
      if (Array.isArray(collection)) {
        // you can use the "collection" to know what was the last array used by the Filter/Editor
        // so if we add the new value to the collection
        collection.push({ value: lastRowIndex, label: lastRowIndex });

        // trigger a change for the Observable/Subject of the async collection
        // I wish that I could get rid of this line and only watch the "collection" change
        // but for now, that is the way I got it to work
        if (asyncCollection instanceof Subject) {
          asyncCollection.next(collection);
        }
      }
    }
  }
```

If someone knows how to do the same as `$watchCollection` like we had in AngularJS, please let me know and provide a solution on the [SO](https://stackoverflow.com/questions/51370933/how-to-watch-a-collection-change-in-angular-5-within-a-service) question, which got down voted :(
I would really like to get rid of the extra ` asyncCollection.next(collection)` call. 

Also note that I did the proof of concept only locally with `setTimeout` and a `Subject` which is not an `HttpClient` call but it should work the same, see below for a quick demo. Below, I'm adding 2 new rows which will add 2 new options to the filter (500 & 501).

![2018-07-17_19-46-35](https://user-images.githubusercontent.com/643976/42851332-255e2cd2-89fa-11e8-9a53-6fecd32284e2.gif)

**Comment by ghiscoding** [comment]
So this ended up being much harder and longer than I expected for the Editor to work (Filter was simpler). The reason is that SlickGrid uses it's own Editor Factory internally (while with Filter, I use my own Factory so I have control over it) and you cannot simply push to the `collection` as @asdman384 tried to do since SlickGrid won't have latest reference (you need to tell SlickGrid to use a newer version of the Editor). 

Anyhow, I got a proof of concept done and I simply need to cleanup some code. I hope to release this in the coming days or week at latest. 

See below for a demo of both Filters & Editors in action. On page load it uses `HttpClient` Observable to pull data from a JSON file, then once I click on "Add Item", it pushes to the `collection` using an RxJS/Subject.

Here it is (Task 100 and 101 are added dynamically)

![2018-08-07_18-46-23](https://user-images.githubusercontent.com/643976/43806698-0f3e598e-9a73-11e8-8526-2ed37d3441fd.gif)

**Comment by ghiscoding** [comment]
This feature is now available in [version 1.4.0](https://github.com/ghiscoding/Angular-Slickgrid/releases/tag/v1.4.0)

If you haven't yet voted, please do ⭐️ 
Cheers
